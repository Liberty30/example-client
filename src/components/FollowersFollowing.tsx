import React, { useEffect, useState } from "react";
import { Button } from "antd";
import FollowersFollowingUsers from "./FollowersFollowingUsers";
import { useAppSelector } from "../redux/hooks";
import { getProfile } from "../services/sdk";
import { Profile } from "../utilities/types";
import { ListStatus } from "../utilities/enums";

const FollowersFollowing = (): JSX.Element => {
  const graph = useAppSelector((state) => state.user.graph);
  const [selectedListTitle, setSelectedListTitle] = useState<ListStatus>(
    ListStatus.CLOSED
  );

  const [usersList, setUsersList] = useState<Profile[] | any>([]);

  const followingList: Profile[] = [];
  const followersList: Profile[] = [];
  graph?.following.map(async (socialAddress) => {
    followingList.push(await getProfile(socialAddress));
  });
  graph?.followers.map(async (socialAddress) => {
    followersList.push(await getProfile(socialAddress));
  });

  const notFollowing = followersList.filter((userProfile) => {
    return !followingList.includes(userProfile);
  });

  useEffect(() => {
    if (selectedListTitle === ListStatus.FOLLOWING) {
      setUsersList(followingList);
    } else if (selectedListTitle === ListStatus.FOLLOWERS) {
      setUsersList(followersList);
    } else setUsersList([]);
  }, [selectedListTitle]);

  const handleClick = (listTitle: ListStatus) => {
    if (selectedListTitle === listTitle)
      setSelectedListTitle(ListStatus.CLOSED);
    else {
      setSelectedListTitle(listTitle);
    }
  };

  const getClassName = (name: ListStatus) => {
    return selectedListTitle === name
      ? "FollowersFollowing__button FollowersFollowing__button--active"
      : "FollowersFollowing__button";
  };

  return (
    <div className="FollowersFollowing__block">
      <div className="FollowersFollowing__buttonBlock">
        <Button
          className={getClassName(ListStatus.FOLLOWERS)}
          onClick={() => handleClick(ListStatus.FOLLOWERS)}
        >
          <div className="FollowersFollowing__buttonCount">
            {graph?.followers.length}
          </div>
          Followers
        </Button>
        {selectedListTitle === ListStatus.CLOSED && (
          <div className="FollowersFollowing__buttonSeperator"> </div>
        )}
        <Button
          className={getClassName(ListStatus.FOLLOWING)}
          onClick={() => handleClick(ListStatus.FOLLOWING)}
        >
          <div className="FollowersFollowing__buttonCount">
            {graph?.following.length}
          </div>
          Following
        </Button>
      </div>
      <FollowersFollowingUsers
        listStatus={selectedListTitle}
        tempUserList={usersList}
        notFollowing={notFollowing}
      />
    </div>
  );
};

export default FollowersFollowing;
