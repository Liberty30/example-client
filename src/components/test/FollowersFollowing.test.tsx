import FollowersFollowing from "../FollowersFollowing";
import { mount } from "enzyme";
import { forcePromiseResolve } from "../../test/testhelpers";
import { componentWithStore, createMockStore } from "../../test/testhelpers";
import { getPreFabSocialGraph } from "../../test/testGraphs";
import { getPrefabSocialAddress } from "../../test/testAddresses";

const profile = getPrefabSocialAddress(0);
const graph = getPreFabSocialGraph().get(profile);
const userGraph = { user: { graph } };
const store = createMockStore(userGraph);

describe("FollowersFollowing", () => {
  it("renders without crashing", async () => {
    const component = mount(componentWithStore(FollowersFollowing, store));
    await forcePromiseResolve();
    expect(() => component).not.toThrow();
  });

  describe("button displays correct list title", () => {
    it("displays correct list title on followers click", async () => {
      const component = mount(componentWithStore(FollowersFollowing, store));
      await forcePromiseResolve();
      component.find(".FollowersFollowing__button").first().simulate("click");
      expect(component.find("FollowersFollowingUsers").prop("listStatus")).toBe(
        1
      );
    });

    it("displays correct list title on following click", async () => {
      const component = mount(componentWithStore(FollowersFollowing, store));
      await forcePromiseResolve();
      component.find(".FollowersFollowing__button").last().simulate("click");
      expect(component.find("FollowersFollowingUsers").prop("listStatus")).toBe(
        2
      );
    });

    it("hides list on double click", async () => {
      const component = mount(componentWithStore(FollowersFollowing, store));
      await forcePromiseResolve();
      component.find(".FollowersFollowing__button").last().simulate("click");
      component.find(".FollowersFollowing__button").last().simulate("click");
      expect(component.find("FollowersFollowingUsers").prop("listStatus")).toBe(
        0
      );
    });

    it("switches back and forth between lists", async () => {
      const component = mount(componentWithStore(FollowersFollowing, store));
      await forcePromiseResolve();
      component.find(".FollowersFollowing__button").first().simulate("click");
      component.find(".FollowersFollowing__button").last().simulate("click");
      expect(component.find("FollowersFollowingUsers").prop("listStatus")).toBe(
        2
      );
    });
  });
});
