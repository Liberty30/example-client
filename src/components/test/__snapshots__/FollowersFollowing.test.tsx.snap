// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FollowersFollowing displays correct list title 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <FollowersFollowing>
        <div
          className="FollowersFollowing__block"
        >
          <div
            className="FollowersFollowing__buttonBlock"
          >
            <Button
              className="FollowersFollowing__button FollowersFollowing__button--active"
              onClick={[Function]}
            >
              <Wave>
                <button
                  className="ant-btn FollowersFollowing__button FollowersFollowing__button--active"
                  onClick={[Function]}
                  type="button"
                >
                  <LoadingIcon
                    existIcon={false}
                    loading={false}
                    prefixCls="ant-btn"
                  >
                    <CSSMotion
                      motionName="ant-btn-loading-icon-motion"
                      onAppearActive={[Function]}
                      onAppearStart={[Function]}
                      onEnterActive={[Function]}
                      onEnterStart={[Function]}
                      onLeaveActive={[Function]}
                      onLeaveStart={[Function]}
                      removeOnLeave={true}
                      visible={false}
                    >
                      <DomWrapper />
                    </CSSMotion>
                  </LoadingIcon>
                  <div
                    className="FollowersFollowing__buttonCount"
                    key=".0"
                  >
                    2
                  </div>
                  <span
                    key=".1"
                  >
                    Followers
                  </span>
                </button>
              </Wave>
            </Button>
            <Button
              className="FollowersFollowing__button"
              onClick={[Function]}
            >
              <Wave>
                <button
                  className="ant-btn FollowersFollowing__button"
                  onClick={[Function]}
                  type="button"
                >
                  <LoadingIcon
                    existIcon={false}
                    loading={false}
                    prefixCls="ant-btn"
                  >
                    <CSSMotion
                      motionName="ant-btn-loading-icon-motion"
                      onAppearActive={[Function]}
                      onAppearStart={[Function]}
                      onEnterActive={[Function]}
                      onEnterStart={[Function]}
                      onLeaveActive={[Function]}
                      onLeaveStart={[Function]}
                      removeOnLeave={true}
                      visible={false}
                    >
                      <DomWrapper />
                    </CSSMotion>
                  </LoadingIcon>
                  <div
                    className="FollowersFollowing__buttonCount"
                    key=".0"
                  >
                    2
                  </div>
                  <span
                    key=".1"
                  >
                    Following
                  </span>
                </button>
              </Wave>
            </Button>
          </div>
          <FollowersFollowingUsers
            listStatus={1}
            notFollowing={
              Array [
                Object {
                  "@context": "https://www.w3.org/ns/activitystreams",
                  "actor": "0x1DCODE2222",
                  "discoverable": true,
                  "icon": Object {
                    "url": "",
                  },
                  "id": "http://localhost:3003/api/announce/0x1DCODE2222",
                  "name": "",
                  "preferredUsername": "",
                  "socialAddress": "0x1DCODE2222",
                  "summary": "",
                  "type": "Person",
                  "url": "",
                  "walletAddress": "0x1DCODE2222",
                },
              ]
            }
            tempUserList={
              Array [
                Object {
                  "@context": "https://www.w3.org/ns/activitystreams",
                  "actor": "0x1DCODE1111",
                  "discoverable": true,
                  "icon": Object {
                    "url": "",
                  },
                  "id": "http://localhost:3003/api/announce/0x1DCODE1111",
                  "name": "",
                  "preferredUsername": "",
                  "socialAddress": "0x1DCODE1111",
                  "summary": "",
                  "type": "Person",
                  "url": "",
                  "walletAddress": "0x1DCODE1111",
                },
                Object {
                  "@context": "https://www.w3.org/ns/activitystreams",
                  "actor": "0x1DCODE2222",
                  "discoverable": true,
                  "icon": Object {
                    "url": "",
                  },
                  "id": "http://localhost:3003/api/announce/0x1DCODE2222",
                  "name": "",
                  "preferredUsername": "",
                  "socialAddress": "0x1DCODE2222",
                  "summary": "",
                  "type": "Person",
                  "url": "",
                  "walletAddress": "0x1DCODE2222",
                },
              ]
            }
          >
            <div
              className="FollowersFollowingUsers__user"
              key="0x1DCODE1111"
            >
              <UserAvatar
                avatarSize="small"
                profile={
                  Object {
                    "@context": "https://www.w3.org/ns/activitystreams",
                    "actor": "0x1DCODE1111",
                    "discoverable": true,
                    "icon": Object {
                      "url": "",
                    },
                    "id": "http://localhost:3003/api/announce/0x1DCODE1111",
                    "name": "",
                    "preferredUsername": "",
                    "socialAddress": "0x1DCODE1111",
                    "summary": "",
                    "type": "Person",
                    "url": "",
                    "walletAddress": "0x1DCODE1111",
                  }
                }
              >
                <Avatar
                  alt="anonymous"
                  shape="circle"
                  size={32}
                  src="data:image/png;base64,00"
                >
                  <span
                    className="ant-avatar ant-avatar-circle ant-avatar-image"
                    style={
                      Object {
                        "fontSize": 18,
                        "height": 32,
                        "lineHeight": "32px",
                        "width": 32,
                      }
                    }
                  >
                    <img
                      alt="anonymous"
                      onError={[Function]}
                      src="data:image/png;base64,00"
                    />
                  </span>
                </Avatar>
              </UserAvatar>
              <div
                className="FollowersFollowingUsers__name"
              >
                Anonymous
              </div>
              <Button
                className="FollowersFollowingUsers__button"
                name="Unfollow"
              >
                <Wave>
                  <button
                    className="ant-btn FollowersFollowingUsers__button"
                    name="Unfollow"
                    onClick={[Function]}
                    type="button"
                  >
                    <LoadingIcon
                      existIcon={false}
                      loading={false}
                      prefixCls="ant-btn"
                    >
                      <CSSMotion
                        motionName="ant-btn-loading-icon-motion"
                        onAppearActive={[Function]}
                        onAppearStart={[Function]}
                        onEnterActive={[Function]}
                        onEnterStart={[Function]}
                        onLeaveActive={[Function]}
                        onLeaveStart={[Function]}
                        removeOnLeave={true}
                        visible={false}
                      >
                        <DomWrapper />
                      </CSSMotion>
                    </LoadingIcon>
                    <span
                      key=".0"
                    >
                      Unfollow
                    </span>
                    <div
                      className="FollowersFollowingUsers__buttonUnfollowIcon"
                      key=".1"
                    >
                      âœ•
                    </div>
                  </button>
                </Wave>
              </Button>
            </div>
            <div
              className="FollowersFollowingUsers__user"
              key="0x1DCODE2222"
            >
              <UserAvatar
                avatarSize="small"
                profile={
                  Object {
                    "@context": "https://www.w3.org/ns/activitystreams",
                    "actor": "0x1DCODE2222",
                    "discoverable": true,
                    "icon": Object {
                      "url": "",
                    },
                    "id": "http://localhost:3003/api/announce/0x1DCODE2222",
                    "name": "",
                    "preferredUsername": "",
                    "socialAddress": "0x1DCODE2222",
                    "summary": "",
                    "type": "Person",
                    "url": "",
                    "walletAddress": "0x1DCODE2222",
                  }
                }
              >
                <Avatar
                  alt="anonymous"
                  shape="circle"
                  size={32}
                  src="data:image/png;base64,00"
                >
                  <span
                    className="ant-avatar ant-avatar-circle ant-avatar-image"
                    style={
                      Object {
                        "fontSize": 18,
                        "height": 32,
                        "lineHeight": "32px",
                        "width": 32,
                      }
                    }
                  >
                    <img
                      alt="anonymous"
                      onError={[Function]}
                      src="data:image/png;base64,00"
                    />
                  </span>
                </Avatar>
              </UserAvatar>
              <div
                className="FollowersFollowingUsers__name"
              >
                Anonymous
              </div>
              <Button
                className="FollowersFollowingUsers__button"
                name="Follow"
              >
                <Wave>
                  <button
                    className="ant-btn FollowersFollowingUsers__button"
                    name="Follow"
                    onClick={[Function]}
                    type="button"
                  >
                    <LoadingIcon
                      existIcon={false}
                      loading={false}
                      prefixCls="ant-btn"
                    >
                      <CSSMotion
                        motionName="ant-btn-loading-icon-motion"
                        onAppearActive={[Function]}
                        onAppearStart={[Function]}
                        onEnterActive={[Function]}
                        onEnterStart={[Function]}
                        onLeaveActive={[Function]}
                        onLeaveStart={[Function]}
                        removeOnLeave={true}
                        visible={false}
                      >
                        <DomWrapper />
                      </CSSMotion>
                    </LoadingIcon>
                    <span
                      key=".0"
                    >
                      Follow
                    </span>
                    <div
                      className="FollowersFollowingUsers__buttonUnfollowIcon"
                      key=".1"
                    >
                      âœ•
                    </div>
                  </button>
                </Wave>
              </Button>
            </div>
          </FollowersFollowingUsers>
        </div>
      </FollowersFollowing>
    </Router>
  </BrowserRouter>
</Provider>
`;
